# Outputs

Analysis workflow results for all samples in the run will appear in the output directory defined by `--outDir` (`out` by default). 

| Path                        | File                                            | Description                                                                                                                                                                                                                                                                                  |
|-----------------------------|-------------------------------------------------|----------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------------|
| `report/`                   | `<sample>.report.html`                          | An interactive standalone HTML report including key metrics/figures <br> for each sample                                                                                                                                                                                                     |
| `fastq/`                    | `<fastqName>.fastq.gz`                          | Fastqs derived from specified `runFolder`  doesn't exist if `fastqDir` passed instead.                                                                                                                                                                                                       |
|                             | `multiqc_report.html`                           | A [multiQC](https://github.com/MultiQC/MultiQC) report containing BCL convert & [fastqc](https://github.com/s-andrews/FastQC) output for the given fastq  |
| `demux/<fastqName>.demux/`  | `<sample>.fastq.gz`                             | Demultiplexed fastq files                                                                                                                                                                                                                                                                    |
|                             | `metrics.json`                                  | Read aligning and Barcode  matching statistics for Samples within `fastqName`. Barcode  statistics separated into: <br> - droplet (drop) <br> - index <br> - tagmentation (tgmt) <br>                                                                                                      |
| `align/`                    | `<sample>.fragments.tsv.gz`                     | Fragments file generated for `<sample>`                                                                                                                                                                                                                                                      |
|                             | `<sample>.bam`                                  | Alignment of reads derived from `<sample>` to  `genome` specified in `experiment.config`                                                                                                                                                                                                     |
| `align/dedup/`              | `<sample>.dedup_stats.tsv`                      | Table denoting proportion of Passing, Unique  and Mitchondrial reads between all cells in  `<sample>`                                                                                                                                                                                        |
|                             | `<sample>.cell_stats.tsv`                       | Table containing number of Passing, Unique  and Mitochondrial reads per barcode.                                                                                                                                                                                                             |
| `peaks/`                    | `<sample>_peaks.narrowPeak`                     | [macs3](https://github.com/macs3-project/MACS) called peaks for `<sample>` in <br> [narrowPeak format](https://macs3-project.github.io/MACS/docs/callpeak.html)                                                                                                                              |
| `peaks/<sample>.counts/`    | `features.tsv`<br>`matrix.mtx`<br>`metrics.tsv` | Components of a **Cell by Peak** count matrix in [MEX format]( https://math.nist.gov/MatrixMarket/formats.html?msclkid=b5e11122d09d11eca0f3a80d769adbbe)  for `<sample>`                                                                                                                     |
|                             | `metrics.tsv`                                   | Table denoting: <br> - unique barcodes counted <br> - unique features counted <br> - reads counted <br>  - fraction of total reads <br> ? Fraction reads counts?  included in peaks                                                                                                          |
|                             | `barcodes.stats.tsv`                            | Table denoting: <br> - number of unique peaks<br> - total counts<br> - Fraction of reads in peaks<br> per barcode                                                                                                                                                                            |
| `peaks/<sample>.tss_counts/` | `features.tsv`<br>`matrix.mtx`<br>`metrics.tsv` | Components of a **Cell by Gene** count matrix (based on reads mapping to peaks in   TSS regions). [MEX format](https://math.nist.gov/MatrixMarket/formats.html?msclkid=b5e11122d09d11eca0f3a80d769adbbe)  for `<sample>`                                                                     |
|                             | `metrics.tsv`                                   | Table denoting: <br> - unique barcodes counted <br> - unique features counted <br> - reads counted <br>  - fraction of total reads <br> ? Fraction reads counts?  included in peaks at TSS                                                                                                   |
|                             | `barcodes.stats.tsv`                            | Table denoting: <br>  - number of unique peaks in TSS regions <br>  - total counts<br> - Fraction of reads in peaks at TSS regions<br>  per barcode                                                                                                                                          |
| `QC/<sample>/`              | `thresholds.json`                                | JSON containing thresholds for minimum  read counts (`ReadsThresh`) and minimum reads  in peaks counts (`ReadsInPeaksThresh`) derived  from passed `QC Filtering Params`                                                                                                                     |
|                             | `QC.tsv`                                        | Table denoting which cells are to be included in  downstream analysis (`Pass`) as well as the qc filtering  criteria they failed if any (`Filter`)                                                                                                                                           |
| `ArchR/<sample>/`*           | `Save-ArchR-Project.rds`                        | R data structure used to reopen automated analysis in  [ArchR](https://www.archrproject.com/index.html). <br> Dependent on: <br>  - `ArrowFiles` <br> - `Embeddings` <br> - `ImputeWeights` <br> - `IterativeLSI` <br> Existing in the same folder to successfully instantiate.              |
|                             | `geneScore`                                     | Components of a **Cell by Gene Score** matrix in [MEX format](https://math.nist.gov/MatrixMarket/formats.html?msclkid=b5e11122d09d11eca0f3a80d769adbbe) .<br> [How ArchR calculates gene scores from ATAC data](https://www.archrproject.com/bookdown/calculating-gene-scores-in-archr.html) |
|                             | `Plots`                                         | PDF files with figures of: <br>- TileMatrix based UMAP <br> - Heatmap of marker genes (by genescore) for all clusters <br> - UMAP embeddings colored by top 10 marker genes for each cluster ** <br>                                                                                  |
|                             | `Top10Markers.tsv`                              | Table denoting (at most) top 10 marker genes (by gene score) for each derived **cluster                                                                                                                                                                                                        |


* *ArchR output is only generated if json specified as the `genome` field in params contains either:
    * **To use builtin annotations**: An `archrAlias`  
            * Must also specify: `params.useBuiltInGenome = true` in `experiment.config`
    * **To use custom annotations**: A name of a BSGenome (as `BSGenome`), a `gtf` and an `annotationConvention`

* **Output related to marker genes is only generated if `getMarkerGenes` is set to `true` in your [qcAndArchRParams yml](../references/parameters/qcAndArchR.yml). Use [this documentation](additionalInputParams.md)
 to modify parameters in this yml.  
